src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/trait/BitbucketPullRequestDiscoveryTrait.java

```diff
@@ -0,0 +1,140 @@
+import jenkins.scm.api.SCMHeadCategory;
+import jenkins.scm.api.SCMRevision;
+import jenkins.scm.api.SCMSource;
+import jenkins.scm.api.trait.SCMBuilder;
+import jenkins.scm.api.trait.SCMSourceContext;
+import jenkins.scm.api.trait.SCMSourceTraitDescriptor;
+import jenkins.scm.impl.ChangeRequestSCMHeadCategory;
+import jenkins.scm.impl.trait.Discovery;
+import org.kohsuke.stapler.DataBoundConstructor;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nullable;
+import javax.inject.Inject;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+public class BitbucketPullRequestDiscoveryTrait extends BitbucketSCMSourceTrait {
+
+    @DataBoundConstructor
+    public BitbucketPullRequestDiscoveryTrait() {
+    }
+
+    @Override
+    protected void decorateContext(SCMSourceContext<?, ?> context) {
+        if (context instanceof BitbucketSCMSourceContext) {
+            BitbucketSCMSourceContext bitbucketContext = (BitbucketSCMSourceContext) context;
+            BitbucketSCMRepository repository = bitbucketContext.getRepository();
+            bitbucketContext.withDiscoveryHandler(
+                    BitbucketDiscoverableHeadType.PULL_REQUEST,
+                    new BitbucketSCMHeadDiscoveryHandler() {
+
+                        @Override
+                        public Stream<SCMHead> discoverHeads() {
+                            DescriptorImpl descriptor = (DescriptorImpl) getDescriptor();
+                            Optional<BitbucketServerConfiguration> mayBeServerConf =
+                                    descriptor.getConfiguration(bitbucketContext.getRepository().getServerId());
+                            if (!mayBeServerConf.isPresent()) {
+                                return Stream.empty();
+                            }
+
+                            BitbucketServerConfiguration serverConfiguration = mayBeServerConf.get();
+                            BitbucketScmHelper scmHelper =
+                                    descriptor.getBitbucketScmHelper(serverConfiguration.getBaseUrl(),
+                                            bitbucketContext.getCredentials().orElse(null));
+
+                            return scmHelper.getOpenPullRequests(repository.getProjectKey(),
+                                            repository.getRepositorySlug())
+                                    .map(BitbucketPullRequestSCMRevision::fromPullRequest)
+                                    .map(BitbucketPullRequestSCMRevision::getHead);
+                        }
+
+                        @Override
+                        public SCMRevision toRevision(SCMHead head) {
+                            return BitbucketPullRequestSCMRevision
+                                    .fromPullRequestHead((BitbucketPullRequestSCMHead) head);
+                        }
+                    }
+            );
+        }
+    }
+
+    @Override
+    protected void decorateBuilder(SCMBuilder<?, ?> builder) {
```

Comment:
Having this decoration here in the trait is an improvement of the [previous spike](https://github.com/jaguilar-atl/atlassian-bitbucket-server-integration-plugin/pull/1/files#diff-933382a481d32a348aa5d3c47a2d3e7b63b23d1189c71f8f66f5d1ce98f0703dR63) as it keeps PR related concerns part of this trait so we don't have to pollute the other classes with it.